#!/bin/bash
# This script was generated by BRB-SeqTools 1.0 (Oct 2016). BRB-SeqTools does not take the responsibility for any damage, loss of data resulting from the use of the software.

set -e
export seqtools_samtools_PATH=/opt/SeqTools/bin/samtools-1.3:/opt/SeqTools/bin/samtools-1.3/misc
export PATH=$seqtools_samtools_PATH:$PATH
export seqtools_bwa_PATH=/opt/SeqTools/bin/bwa-0.7.12
export PATH=$seqtools_bwa_PATH:$PATH

export seqtools_bcftools_PATH=/opt/SeqTools/bin/bcftools-1.3
export PATH=$seqtools_bcftools_PATH:$PATH

export seqtools_GATK_PATH=/opt/SeqTools/bin/gatk
export PATH=$seqtools_GATK_PATH:$PATH

outputDir="/home/brb//testdata/GSE48215subset/outputGatk"
if [ -f "$outputDir"/BugReport.tar.gz ]; then rm "$outputDir"/BugReport.tar.gz; fi;
if ls "$outputDir"/*.count &> /dev/null; then rm "$outputDir"/*.count ; fi
if ls "$outputDir"/*.vcf &> /dev/null; then rm "$outputDir"/*.vcf ; fi
if ls "$outputDir"/*.idx &> /dev/null; then rm "$outputDir"/*.idx ; fi
if [ -d "$outputDir/tmp" ]; then rm -r -f "$outputDir/tmp"; fi

if [ ! -d "$outputDir/tmp" ]; then mkdir "$outputDir/tmp"; fi

if [ -f "$outputDir/tmp/log.txt" ]; then rm "$outputDir/tmp/log.txt"; fi
touch "$outputDir/tmp/log.txt"

echo bowtie: /opt/SeqTools/bin/bowtie2-2.2.6 >> "$outputDir/tmp/log.txt"
echo tophat: /opt/SeqTools/bin/tophat-2.1.0.Linux_x86_64 >> "$outputDir/tmp/log.txt"
echo samtools: /opt/SeqTools/bin/samtools-1.3 >> "$outputDir/tmp/log.txt"
echo HTSeq: `python -c "import HTSeq; print HTSeq.__version__"` >> "$outputDir/tmp/log.txt"

runDGE=false
runVC=true
rerunAligner=false
cleanUp=true

cd "/home/brb//testdata/GSE48215subset"

# bt20 
echo bt20 `date +'%Y-%m-%d %T'` >> "$outputDir/tmp/log.txt"
echo "bt20 (1/1)"
if [ ! -d "bt20" ]; then mkdir "bt20"; fi

    if [ ! -f bt20/accepted_hits_bwa.bam ] || ([ -f bt20/accepted_hits_bwa.bam ] && [ $rerunAligner == "true" ]) ; then
        # MappingWithBWA. Input: fastq. Output: accepted_hits_bwa.bam
        echo "    running read alignment" | tee -a "$outputDir/tmp/log.txt"
        (bwa mem  -t 1 -M "/home/brb//testdata/hg19/chr1.fa" SRR925751_1.fastq SRR925751_2.fastq > "bt20/bwa.sam") >> "$outputDir/tmp/log.txt" 2>&1
        (samtools fixmate -O bam "bt20/bwa.sam" "bt20/accepted_hits_bwa.bam") >> "$outputDir/tmp/log.txt" 2>&1
    fi;
if [ $runDGE == "true" ]; then
    cd "bt20"
    echo "    sorting bam file" | tee -a "$outputDir/tmp/log.txt"
    (samtools sort -@ 1 -n accepted_hits_bwa.bam -o sortName.bam) >> "$outputDir/tmp/log.txt" 2>&1
    # samtools view -o sortName.sam sorted.bam
    echo "    creating count file" | tee -a "$outputDir/tmp/log.txt"
    (python -m HTSeq.scripts.count -f bam -s no -a 10 sortName.bam "/home/brb//testdata/hg19/chr1.gtf" > "/home/brb//testdata/GSE48215subset/outputGatk/bt20.count") >> "$outputDir/tmp/log.txt" 2>&1
    if [ $cleanUp == "true" ]; then
        rm -rf `ls | grep -v 'accepted_hits_tophat.bam' | grep -v 'accepted_hits_star.bam' | grep -v 'accepted_hits_bwa.bam' | grep -v 'accepted_hits_usr.bam' | grep -v 'align_summary.txt' | grep -v 'Log.final.out' `
    fi;
    cd ..
fi

if [ $runVC == "true" ]; then
    # SortAndIndexBam. Input: accepted_hits.bam. Output: sorted.bam
    cd "bt20"
    echo "    sorting bam file" >> "$outputDir/tmp/log.txt"

    (samtools sort -@ 1 accepted_hits_bwa.bam -o sorted.bam) >> "$outputDir/tmp/log.txt" 2>&1
    echo "    indexing bam file" >> "$outputDir/tmp/log.txt"

    (samtools index sorted.bam) >> "$outputDir/tmp/log.txt" 2>&1
    # MarkDuplication (optional). Input: sort.bam. Output: Libraryname.bam
    echo "    marking duplicates" | tee -a "$outputDir/tmp/log.txt"
    (java -Xmx10g -jar /opt/SeqTools/bin/picard-tools-1.141/picard.jar  MarkDuplicates METRICS_FILE=MarkDudup.metrics INPUT=sorted.bam OUTPUT=bt20.bam) >> "$outputDir/tmp/log.txt" 2>&1
    echo "    creating .intervals file for running the GATK pipeline" | tee -a "$outputDir/tmp/log.txt"
    (java -Xmx10g -jar /opt/SeqTools/bin/gatk/GenomeAnalysisTK.jar -T RealignerTargetCreator -R /home/brb//testdata/hg19/chr1.fa -o /home/brb/testdata/hg19/BRB-SeqTools_indels_only.intervals -known /home/brb//testdata/hg19/common_all_20160601_1.vcf.gz -nt 1) >> "$outputDir/tmp/log.txt" 2>&1
    #Assign read group information. 
    (java -Xmx10g -jar /opt/SeqTools/bin/picard-tools-1.141/picard.jar AddOrReplaceReadGroups INPUT=bt20.bam OUTPUT=rg_added_sorted.bam RGID=1 RGLB=dna RGPL=illumina RGPU=UNKNOWN RGSM=bt20) >> "$outputDir/tmp/log.txt" 2>&1
    #Reorder sam file.
    (java -Xmx10g -jar /opt/SeqTools/bin/picard-tools-1.141/picard.jar  ReorderSam INPUT=rg_added_sorted.bam OUTPUT=reorder.bam REFERENCE=/home/brb//testdata/hg19/chr1.fa) >> "$outputDir/tmp/log.txt" 2>&1
    mv ./reorder.bam ./split.bam
    samtools index split.bam
    echo "    running indel realignment" | tee -a "$outputDir/tmp/log.txt"
    (java -Xmx10g -jar /opt/SeqTools/bin/gatk/GenomeAnalysisTK.jar -T IndelRealigner -R /home/brb//testdata/hg19/chr1.fa -I split.bam -targetIntervals /home/brb/testdata/hg19/BRB-SeqTools_indels_only.intervals -known /home/brb//testdata/hg19/common_all_20160601_1.vcf.gz -o realigned_reads.bam -allowPotentiallyMisencodedQuals) >> "$outputDir/tmp/log.txt" 2>&1
    echo "    base quality score recalibration" | tee -a "$outputDir/tmp/log.txt"
    (java -Xmx10g -jar /opt/SeqTools/bin/gatk/GenomeAnalysisTK.jar -T BaseRecalibrator -R /home/brb//testdata/hg19/chr1.fa -I realigned_reads.bam -nct 1 -knownSites /home/brb//testdata/hg19/common_all_20160601_1.vcf.gz -o recal_data.table -allowPotentiallyMisencodedQuals) >> "$outputDir/tmp/log.txt" 2>&1
    (java -Xmx10g -jar /opt/SeqTools/bin/gatk/GenomeAnalysisTK.jar -T PrintReads -R /home/brb//testdata/hg19/chr1.fa -I realigned_reads.bam -nct 1 -BQSR recal_data.table -o recal.bam) >> "$outputDir/tmp/log.txt" 2>&1
    echo "    calling variants" | tee -a "$outputDir/tmp/log.txt"
    (java -Xmx10g -jar /opt/SeqTools/bin/gatk/GenomeAnalysisTK.jar -T HaplotypeCaller --genotyping_mode DISCOVERY -R /home/brb//testdata/hg19/chr1.fa -I recal.bam  -stand_call_conf 0 -stand_emit_conf 0 -o /home/brb//testdata/GSE48215subset/outputGatk/bt20_raw.vcf -allowPotentiallyMisencodedQuals -nct 1) >> "$outputDir/tmp/log.txt" 2>&1
    if [ $cleanUp == "true" ]; then
        rm -rf `ls | grep -v 'accepted_hits_tophat.bam' | grep -v 'accepted_hits_star.bam' | grep -v 'accepted_hits_bwa.bam' | grep -v 'accepted_hits_usr.bam'  | grep -v 'align_summary.txt'`
    fi;
    cd ..
    echo "    finish variant call" | tee -a "$outputDir/tmp/log.txt"
    # echo "    bt20_raw.vcf" is created | tee -a "$outputDir/tmp/log.txt"
fi

echo completed `date +'%Y-%m-%d %T'` >> "$outputDir/tmp/log.txt"
